pkg/rootless/rootless_linux.c

```diff
@@ -293,6 +301,24 @@ static void __attribute__((constructor)) init()
       closedir (d);
     }
 
+    listen_pid = getenv("LISTEN_PID");
+    listen_fds = getenv("LISTEN_FDS");
+    listen_fdnames = getenv("LISTEN_FDNAMES");
+
+    if (listen_pid != NULL && listen_fds != NULL && strtol(listen_pid, NULL, 10) == getpid())
+      {
+        size_t len_pid = strlen(listen_pid) + 1;
+        size_t len_fds = strlen(listen_fds) + 1;
+        size_t len_fdnames = strlen(listen_fdnames) + 1;
+
+        // save systemd socket environment for rootless child
+        do_socket_activation = true;
+        char * save_buf = (char *) calloc (len_pid + len_fds + len_fdnames, 1);
```

Comment:
can we add a check for `calloc` failures?

Reply:
also there is no need to use a single buffer here.  Let's split it to multiple assignments and use `strdup` instead. Each of the `strdup` must be checked for errors.